### PR TITLE
RN114 validation verifying RN headers converting to the new format

### PR DESCRIPTION
**Pull request by maimorag**
## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-10800

## Description

converting RN114 to the new format

### CODE REVIEW

**Review by dorschw — COMMENTED** [review]
Nice! See the comments

File: demisto_sdk/commands/common/constants.py [hunk 1]
```diff
@@ -2202,3 +2202,10 @@ class PlaybookTaskType(StrEnum):
 INVALID_IMAGE_PATH_REGEX = (
     r"(\!\[.*?\]|src\=)(\(|\")(https://github.com/demisto/content/blob/.*?)(\)|\")"
 )
+# Release note regex
+CONTENT_TYPE_SECTION_REGEX = (
+    r"^#### ([\w ]+)$\n([\w\W]*?)(?=^#### )|^#### ([\w ]+)$\n([\w\W]*)"
+)
+CONTENT_ITEM_SECTION_REGEX = (
+    r"^##### (.+)$\n([\w\W]*?)(?=^##### )|^##### (.+)$\n([\w\W]*)|" r"^- (?:New: )?$"
+)
```
[the remark this answers is not in the document]
Comment (maimorag) [reply demisto_sdk/commands/common/constants.py:2211]: release_note.py

File: demisto_sdk/commands/content_graph/common.py [hunk 2]
```diff
@@ -277,6 +278,43 @@ def by_schema(path: Path) -> "ContentType":
         raise ValueError(f"Could not find content type in path {path}")
 
 
+RN_HEADER_BY_CONTENT_TYPE = {
```
[the remark this answers is not in the document]
Comment (maimorag) [reply demisto_sdk/commands/content_graph/common.py:281]: to the RN validation? if so, no problem

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    ContentType.GENERIC_TYPE: "Object Types",
+    ContentType.GENERIC_FIELD: "Object Fields",
+    ContentType.LIST: "Lists",
+    ContentType.JOB: "Jobs",
+    ContentType.PARSING_RULE: "Parsing Rules",
+    ContentType.MODELING_RULE: "Modeling Rules",
+    ContentType.CORRELATION_RULE: "Correlation Rules",
+    ContentType.XSIAM_DASHBOARD: "XSIAM Dashboards",
+    ContentType.XSIAM_REPORT: "XSIAM Reports",
+    ContentType.TRIGGER: "Triggers Recommendations",  # https://github.com/demisto/etc/issues/48153#issuecomment-1111988526
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:49]: fantastic use of a comment!

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    header: content_type for content_type, header in RN_HEADER_BY_CONTENT_TYPE.items()
+}
+
+
+class ReleaseNoteHeaderValidator(BaseValidator[ContentTypes]):
+    error_code = "RN114"
+    description = (
+        "Validate the 2nd headers (the content items) are exists in the pack and having the right display"
+        " name."
+    )
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:68]:
```suggestion
    description = (
        "Validate the content items mentioned in second-level headers (##), exist in the pack."
    )
```

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    rationale = (
+        "Provide documentation with clear headers for all modifications to make content usage easier."
+        " Validate headers for accuracy."
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:71]:
```suggestion
        "Providing documentation with accurate headers, and avoid confusion."
```

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    )
+    error_message = (
+        "The following invalid headers were found in:\n"
+        "{content_type_message}\n{content_item_message}\n"
+        "For common troubleshooting steps, please review the documentation found here: "
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:76]:
```suggestion
        "For common troubleshooting steps, see "
```

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+        "https://xsoar.pan.dev/docs/integrations/changelog#common-troubleshooting-tips"
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:77]: I don't think that in this particular validation, there's much value sending users to that page, better remove the link imo.

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    )
+    related_field = "release_note"
+    is_auto_fixable = False
+    related_file_type = [RelatedFileType.RELEASE_NOTE]
+
+    def is_valid(self, content_items: Iterable[ContentTypes]) -> List[ValidationResult]:
+        validator_results: List[ValidationResult] = []
+        for content_item in content_items:
+            if content_item.pack_metadata_dict and content_item.pack_metadata_dict.get(
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:86]: when would there not be a `content_item.pack_metadata_dict`?

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+                "hidden"
+            ):
+                continue
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:89]: please add a debug log

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+            (
+                invalid_headers_content_type,
+                invalid_headers_content_item,
+            ) = self.validate_release_notes_headers(content_item)
+            content_type_message = (
+                "Headers Content Types: {}\n".format(
+                    ", ".join(invalid_headers_content_type)
+                )
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:97]:
```suggestion
                f"Headers Content Types: {', '.join(invalid_headers_content_type)}\n"
                )
```

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+                if invalid_headers_content_type
+                else ""
+            )
+            content_item_message = (
+                "Headers Content Items: {}\n".format(
+                    ", ".join(invalid_headers_content_item)
+                )
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:104]:
```suggestion
                f"Headers Content Items: {', '.join(invalid_headers_content_item)}\n"
                )
```

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        """
+        Filters out None values from a list or tuple.
+
+        Args:
+            ls (List or Tuple): The list or tuple to filter.
+
+        Returns:
+            List: Filtered list with None values removed.
+        """
+        return list(filter(None, ls))
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:131]: doesn't use `self`, better move it out of the class

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+        headers: Dict[str, List[str]] = {}
+        content_type_section_pattern = re.compile(CONTENT_TYPE_SECTION_REGEX, re.M)
+        content_item_section_pattern = re.compile(CONTENT_ITEM_SECTION_REGEX, re.M)
+
+        # Get all sections from the release notes using regex
+        rn_sections = content_type_section_pattern.findall(release_note_content)
+        for section in rn_sections:
+            section = self.remove_none_values(ls=section)
+            if not section:
+                continue
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:153]: add a debug log saying that we ignore an empty section

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+                # Will raise error in rn_valid_header_format.
+                headers[content_type] = []
+
+            for content_type_section in content_type_sections_ls:
+                content_type_section = self.remove_none_values(ls=content_type_section)
+                if content_type_section:
+                    header = (
+                        content_type_section[0].rstrip()
+                        if ("New: " not in content_type_section[0])
+                        else content_type_section[0][5:]
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:172]: why 5? worth a comment

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+                    )
+                    if content_type in headers:
+                        headers[content_type].append(header)
+                    else:
+                        headers[content_type] = [header]
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:177]: this can be much clearer using `headers = defaultdict(list)`, and then always appending (even if the key doesn't exist, defaultdict handles it).

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+        Returns:
+            A tuple containing two lists:
+            - List of invalid header types (str or None).
+            - List of invalid header content items (str or None).
+        """
+        headers = self.extract_rn_headers(content_item.release_note.file_content)
+        pack_items_by_types = content_item.content_items.items_by_type()
+        invalid_content_type: List[str] = [
+            header_type
+            for header_type in headers.keys()
```
Comment (dorschw) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:239]:
```suggestion
            for header_type in headers
```

**Review by dorschw — COMMENTED** [review]
Great job!

File: demisto_sdk/commands/common/constants.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dorschw) [reply demisto_sdk/commands/common/constants.py:2222]: if it's only used for the validation, then yes

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py (hunk 8 above)
Comment (maimorag) [reply demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:90]: I can add a ignore but it's a pylance error

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py (hunk 13 above)
Comment (maimorag) [reply demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:150]: added

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py (hunk 9 above)
Comment (maimorag) [reply demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:102]: added

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+from demisto_sdk.commands.content_graph.objects.content_item import ContentItem
+from demisto_sdk.commands.content_graph.objects.pack import Pack
+from demisto_sdk.commands.content_graph.parsers.related_files import RelatedFileType
+from demisto_sdk.commands.validate.validators.base_validator import (
+    BaseValidator,
+    ValidationResult,
+)
+
+ContentTypes = Pack
+RN_HEADER_BY_CONTENT_TYPE = {
```
Comment (dantavori) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:22]: instead of maintaining this dict, let's add to ContentType a property method `release_note_header()` that returns something like `pascalToSpace(self.as_folder)`, and handle manually special cases like `Mappers`, `PreProcess Rules`, `Triggers Recommendations`

File: demisto_sdk/commands/validate/tests/RN_validators_test.py [hunk 18]
```diff
@@ -77,3 +81,64 @@ def test_release_note_filled_out_validator(
             for result, expected_msg in zip(results, expected_msgs)
         ]
     )
+
+
+def test_release_note_header_validator_valid():
```
Comment (dantavori) [inline demisto_sdk/commands/validate/tests/RN_validators_test.py:86]: please do also for a mapper and a trigger (edge cases)

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+ContentTypes = Pack
+
+
+CONTENT_TYPE_SECTION = re.compile(r"^#### ([\w ]+)$\n([\w\W]*?)(?=^#### |\Z)", re.M)
+CONTENT_ITEM_SECTION = re.compile(
+    r"^##### (.+)$\n([\w\W]*?)(?=^##### |\Z)|^- (?:New: )?$", re.M
+)
+
+
+def remove_none_values(ls: Union[List[Any], Tuple[Any, ...]]) -> List[Any]:
```
Comment (dantavori) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:30]: don't really need a method for this
x = [i for i in x if i is not None]

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+
+
+class ReleaseNoteHeaderValidator(BaseValidator[ContentTypes]):
+    error_code = "RN114"
+    description = "Validate the content items mentioned in second-level headers (##), exist in the pack."
+    rationale = "Providing documentation with accurate headers, and avoid confusion."
+    error_message = (
+        "The following headers that were found were invalid:\n"
+        "{content_type_message}\n{content_item_message}\n"
+    )
```
Comment (dantavori) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:50]:
```suggestion
    description = "Validate the existence of content types in the first-level headers (####) and the content items in second-level headers (#####)."
    rationale = "Providing documentation with accurate information and avoiding confusion."
    error_message = (
        "The following release note headers are invalid:\n"
        "{content_type_message}\n{content_item_message}\n"
    )
```
correct me if I'm wrong, we don't validate only the content items but also the types

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+                "hidden"
+            ):
+                logger.debug("Content item is marked as hidden. Skipping.")
+                continue
+            (
+                invalid_headers_content_type,
+                invalid_headers_content_item,
+            ) = self.validate_release_notes_headers(content_item)
+            content_type_message = (
+                f"Headers Content Types: {', '.join(invalid_headers_content_type)}\n"
```
Comment (dantavori) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:68]:
```suggestion
                f"Content types: {', '.join(invalid_headers_content_type)}\n"
```

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+                if invalid_headers_content_type
+                else ""
+            )
+            content_item_message = (
+                f"Headers Content Items: {', '.join(invalid_headers_content_item)}\n"
```
Comment (dantavori) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:73]:
```suggestion
                f"Content items: {', '.join(invalid_headers_content_item)}\n"
```

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+            content_type = ContentType.from_release_note_header(header)
+            pack_display_names = {
+                item.display_name for item in pack_items_by_types.get(content_type, [])
+            }
+            if missing := set(display_names).difference(pack_display_names):
+                missing_display_names[header] = missing
+        return missing_display_names
+
+    def validate_release_notes_headers(
+        self, content_item: Pack
```
Comment (dantavori) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:174]:
```suggestion
        self, pack: Pack
```
easier to read since we know it's a pack

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+
+    def validate_content_type_header(self, header: str) -> bool:
+        """
+            Validate that the release notes 1st headers (the content type) are a valid content entity.
+        Args:
+            content_type: (str) - The content type to validate.(e.g. Integrations, Playbooks, etc.)
+        Return:
+            True if the content type is valid, False otherwise.
+        """
+        return header in RN_HEADER_BY_FILE_TYPE.values()
```
Comment (dantavori) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:142]: we want to avoid using this dict

**Review by dantavori — CHANGES_REQUESTED** [review]
nice, see notes

File: demisto_sdk/commands/content_graph/common.py [hunk 25]
```diff
@@ -276,6 +276,81 @@ def by_schema(path: Path, git_sha: Optional[str] = None) -> "ContentType":
...
+    @property
+    def convert_content_type_to_rn_header(self) -> str:
+        """
+        Convert ContentType to the Release note header.
+        """
+        if self == ContentType.PREPROCESS_RULE:
+            return "PreProcess Rules"
+        elif self == ContentType.TRIGGER:
+            return "Triggers Recommendations"  # https://github.com/demisto/etc/issues/48153#issuecomment-1111988526
+        elif self == ContentType.XSIAM_REPORT:
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/common.py:288]: we can do pascalToSpace to many of the types here (need to add this method to tools.py)

File: demisto_sdk/commands/content_graph/common.py [hunk 26]
```diff
@@ -276,6 +276,81 @@ def by_schema(path: Path, git_sha: Optional[str] = None) -> "ContentType":
...
+        elif header == "Modules":
+            return ContentType.GENERIC_MODULE
+        elif header == "Object Types":
+            return ContentType.GENERIC_TYPE
+        elif header == "Object Fields":
+            return ContentType.GENERIC_FIELD
+        normalized_header = header.rstrip("s").replace(" ", "_").upper()
+        if not ContentType.__dict__.get(normalized_header):
+            raise ValueError
+        return ContentType[normalized_header]
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/common.py:352]: equivalent:
```suggestion
        return ContentType(normalized_header)
```

File: demisto_sdk/commands/content_graph/tests/parsers_and_models_test.py [hunk 27]
```diff
@@ -3008,3 +3008,56 @@ def test_get_related_text_file():
...
+        ContentType.TRIGGER,
+        ContentType.WIZARD,
+        ContentType.XDRC_TEMPLATE,
+        ContentType.LAYOUT_RULE,
+        ContentType.ASSETS_MODELING_RULE,
+        ContentType.CASE_LAYOUT_RULE,
+        ContentType.CASE_FIELD,
+        ContentType.CASE_LAYOUT,
+    ]
+    for content_type in test_cases:
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/tests/parsers_and_models_test.py:3060]:
```suggestion
    for content_type in ContentType:
```

File: demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py [hunk 28]
```diff
@@ -0,0 +1,210 @@
...
+        for content_item in content_items:
+            if content_item.pack_metadata_dict and content_item.pack_metadata_dict.get(
+                "hidden"
+            ):
+                logger.debug("Content item is marked as hidden. Skipping.")
+                continue
+            (
+                invalid_headers_content_type,
+                invalid_headers_content_item,
+            ) = self.validate_convert_content_type_to_rn_headers(content_item)
```
Comment (dantavori) [inline demisto_sdk/commands/validate/validators/RN_validators/RN114_validate_release_notes_header.py:54]:
```suggestion
            ) = self.validate_release_notes_headers(content_item)
```
the previous one was ok :)

**Review by dantavori — APPROVED** [review]
nice, approving but see last notes

File: demisto_sdk/commands/content_graph/common.py (hunk 26 above)
Comment (maimorag) [reply demisto_sdk/commands/content_graph/common.py]: ContentType[normalized_header] retrieves an enum member by its name, and ContentType(normalized_header)- Retrieves an enum member by its value, and here I need to return the value by its name so ContentType(normalized_header) doesn't work the same here
